### PR TITLE
feishin: init at 0.2.0

### PR DESCRIPTION
**Pull request by onny**
###### Description of changes

Package [feishin](https://github.com/jeffvli/feishin), "a full-featured Subsonic/Jellyfin compatible desktop music player"

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/feishin/default.nix
Comment (AndersonTorres) [inline pkgs/applications/audio/feishin/default.nix:16]: Is this `rec` needed? Given that all possibly recursive references are `let`-enclosed...

File: pkgs/applications/audio/feishin/default.nix
Comment (AndersonTorres) [inline pkgs/applications/audio/feishin/default.nix:4]: Is this allowed for this particular use case?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by onny** [comment]
Thank you for reviewing, fixed the issues :)

File: pkgs/applications/audio/feishin/default.nix
Comment (AndersonTorres) [inline pkgs/applications/audio/feishin/default.nix:28]:
```suggestion
  extraPkgs = [
```

Comment (onny) [reply pkgs/applications/audio/feishin/default.nix:28]: Getting following error when trying to remove this
```
error: attempt to call something which is not a function but a list

       at /home/onny/projects/nixpkgs/pkgs/build-support/appimage/default.nix:51:49:

           50|       targetPkgs = pkgs: [ appimage-exec ]
           51|         ++ defaultFhsEnvArgs.targetPkgs pkgs ++ extraPkgs pkgs;
             |                                                 ^
           52|
(use '--show-trace' to show detailed location information)
```

**Comment by happysalada** [comment]
It looks like there are merge conflicts and that the ci is unhappy

**Comment by nephalemsec** [comment]
I've also previously packaged this for my own personal use, feel free to steal the darwin logic:

```nix
{ lib, stdenv, fetchurl, makeWrapper, electron_24, makeDesktopItem
, graphicsmagick, unzip, mpv }:
let
  pname = "feishin";
  version = "0.2.0";
  appname = "Feishin";
  meta = with lib; {
    description = "Rewrite of Sonixd.";
    homepage = "http://feishin.vercel.app/";
    downloadPage = "https://github.com/jeffvli/feishin/releases";
    license = licenses.mit;
  };

  filename = if stdenv.isDarwin then
    "${appname}-${version}-mac-x64.zip"
  else
    "${appname}-${version}-linux-x64.tar.xz";
  src = fetchurl {
    url =
      "https://github.com/jeffvli/feishin/releases/download/v${version}/${filename}";
    sha256 = if stdenv.isDarwin then
      "sha256-WzU/Yd3cNMIpmkKWC29mQlviYXiYV5k8+80iyzilPGc="
    else
      "sha256-o+fEjdG2iN84FEchyAKWZCpsrqDm1K1k+Q2K+cwYBZE=";
  };

  icon = fetchurl {
    url =
      "https://github.com/jeffvli/feishin/raw/development/assets/icons/1024x1024.png";
    sha256 = "sha256-QS2dc+Q+CWx27ZKDOvFC/rj8TZ2HkYGgudYZJfyzDt4=";
  };

  desktopItem = makeDesktopItem {
    name = "feishin";
    desktopName = "Feishin";
    comment = "Remake of Sonixd.";
    icon = "feishin";
    exec = "feishin %u";
    categories = [ "Audio" ];
    mimeTypes = [ "x-scheme-handler/feishin" ];
  };

  linux = stdenv.mkDerivation {
    inherit appname pname version src desktopItem icon;
    meta = meta // { platforms = [ "x86_64-linux" "aarch64-linux" ]; };
    nativeBuildInputs = [ makeWrapper graphicsmagick ];

    # Installs mpv as a requirement
    propagatedBuildInputs = [ mpv ];

    installPhase = ''
      runHook preInstall

      mkdir -p $out/bin
      mkdir -p $out/share
      cp -r resources $out/share/${pname}/
      cp -r locales $out/share/${pname}/

      makeWrapper ${electron_24}/bin/electron $out/bin/${pname} \
        --add-flags $out/share/${pname}/app.asar
      install -m 444 -D "${desktopItem}/share/applications/"* \
        -t $out/share/applications/
      for size in 16 24 32 48 64 128 256 512; do
        mkdir -p $out/share/icons/hicolor/"$size"x"$size"/apps
        gm convert -resize "$size"x"$size" ${icon} $out/share/icons/hicolor/"$size"x"$size"/apps/${appname}.png
      done
      runHook postInstall
    '';

    shellHook = ''
      set -x
      export LD_LIBRARY_PATH=${mpv}/lib
      set +x
    '';
  };

  darwin = stdenv.mkDerivation {
    inherit pname version src appname;
    meta = meta // { platforms = [ "x86_64-darwin" "aarch64-darwin" ]; };
    sourceRoot = "${appname}.app";

    nativeBuildInputs = [ makeWrapper unzip ];

    # Installs mpv as a requirement
    propagatedBuildInputs = [ mpv ];

    installPhase = ''
      runHook preInstall
      mkdir -p $out/{Applications/${appname}.app,bin}
      cp -R . $out/Applications/${appname}.app
      makeWrapper $out/Applications/${appname}.app/Contents/MacOS/${appname} $out/bin/${pname}
      runHook postInstall
    '';

    shellHook = ''
      set -x
      export LD_LIBRARY_PATH=${mpv}/lib
      set +x
    '';
  };
in if stdenv.isDarwin then darwin else linux
```

**Comment by nephalemsec** [comment]
Forgive the ignorance. What is holding this PR back from being merged? I can't see details of the ofborg-eval step.

**Comment by onny** [comment]
I have to look into it but can't tell when I'll find time for it. Feel free to open a new PR, I will be happy reviewing it :)

**Comment by onny** [comment]
@nephalemsec Thank you I used your snippet now in this PR. Should I add you as a co-maintainer? Please check if the Darwin package still works for you :)

File: pkgs/applications/audio/feishin/default.nix
Comment (AndersonTorres) [inline pkgs/applications/audio/feishin/default.nix:103]: To be honest, this expression looks unnecessarily complicated. I vote for
- Split it into two files, one for each system
- `default.nix` would call the respective file

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/feishin/default.nix
Comment (minijackson) [reply pkgs/applications/audio/feishin/default.nix:10]: `downloadPage` is a standard meta attribute: https://nixos.org/manual/nixpkgs/stable/#var-meta-downloadPage

**Comment by onny** [comment]
Thanks @AndersonTorres, I splitted the file into one `linux.nix` and one `darwin.nix`. I tested the Linux build and it's working fine. Still somone needs to test the Darwin build :)

File: pkgs/applications/audio/feishin/darwin.nix
Comment (AndersonTorres) [inline pkgs/applications/audio/feishin/darwin.nix:2]: `meta` is not referenced in the code of this file.
Is this expected somehow?

**Review by AndersonTorres — COMMENTED** [review]
Besides, LGTM :)

**Comment by zowoq** [comment]
https://hydra.nixos.org/build/229890242

This broke the tarball job on hydra, should be fixed by https://github.com/NixOS/nixpkgs/pull/246043.

**Comment by zimbatm** [comment]
@onny this is the second PR that you merged with a broken eval (see https://github.com/NixOS/nixpkgs/pull/241299). This is particularly bad as it has an impact on all new contributions as their ofborg-eval will be failing.

I'm expecting committers to take ownership when creating issues for others. That means (1) acknowledging that this is an issue, (2) working actively to fix the issue.

Please give me a sign that you understand this, otherwise, I will be forced to remove your access. Feel free to reach out privately if you have questions.

**Comment by onny** [comment]
@zimbatm Thank you for your feedback, I noticed the mistake and will tale further caution in the future (additional local build and checking ofborg status).
I‘ll try to respond faster if any issue appears in the future and going to solve it as a priority.

**Comment by zimbatm** [comment]
Thanks. I'll also write a guide to set some of those expectations so it's clear for everybody.

**Comment by Gecked-Deck** [comment]
Anyone else having issues with Feishin in the system tray?

**Comment by nephalemsec** [comment]
I have issues in the artist page and with it not playing sometimes without hitting play/pause again. But those are more app specific, not derivation specific.
Is the icon not showing in tray or something?

**Comment by Gecked-Deck** [comment]
Icon is there, but is invisible and non-interactable. Just an empty space.

**Comment by nephalemsec** [comment]
> Icon is there, but is invisible and non-interactable. Just an empty space.

I cant reproduce your problem sorry. Mine works on MacOS (aarch) and NixOS (x86_64)

There was an update to Feishin (0.3.0), maybe that will fix your issues? The new hashes are as follows:

```
  filename = if stdenv.isDarwin then
    "${appname}-${version}-mac-x64.zip"
  else
    "${appname}-${version}-linux-x64.tar.xz";
  src = fetchurl {
    url =
      "https://github.com/jeffvli/feishin/releases/download/v${version}/${filename}";
    sha256 = if stdenv.isDarwin then
      "sha256-J5LB4uR/NJ6ykiTqBY1VepcLujprgqwpxy7sGD0NtZw="
    else
      "sha256-sl2zM24bb0yBTfCxtNGizp6Yu+L4nj/Uf669zylnPmE=";
  };
```

**Comment by onny** [comment]
Update to 0.3.0 already in nixpkgs master. Please give it a try and otherwise open a new issue :)

**Comment by nephalemsec** [comment]
> Update to 0.3.0 already in nixpkgs master. Please give it a try and otherwise open a new issue :)

Hahaha my bad buddy, been away for a few weeks and didn't realise this got merged. Good job!

**Pull request merged**
